### ISSUE TITLE
Why the gradient in backward after the scaled sign function looks unlike the gradient in the paper?

### ISSUE THREAD

**Issue by iridescent303** [jiecaoyu/XNOR-Net-PyTorch#1]
Is there any trick here?

**Comment by jiecaoyu** [comment]
In the original paper, the gradient is calculated by
  
![equation](http://latex.codecogs.com/gif.latex?%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20W_i%7D%3D%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20%7B%5Cwidetilde%7BW%7D%7D_i%7D%20%28%5Cfrac%7B1%7D%7Bn%7D+%5Cfrac%7B%5Cpartial%20sign%28W_i%29%7D%7B%5Cpartial%20W_i%7D%5Ccdot%20%5Calpha%20%29)

which is got from

![equation](https://www.latex4technics.com/l4ttemp/gk45b3.png?1506583668797)

But it actually should be

![equation](https://www.latex4technics.com/l4ttemp/gk45b3.png?1506583574793)

Therefore, I correct it in my code.

**Comment by iridescent303** [comment]
Do you mean the gradient expression isn't 
![image](https://user-images.githubusercontent.com/16323489/30955774-61f5c584-a467-11e7-9387-f12cc61a0acb.png)? Could you please tell me the gradient expression that you refer to?

**Comment by jiecaoyu** [comment]
Hi, I am using:

![equation](http://latex.codecogs.com/gif.latex?%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20W_%7Bi%7D%7D%3D%5Cfrac%7B1%7D%7Bn%7D%5Ccdot%20sign%28W_%7Bi%7D%29%5Ccdot%20%5Csum_%7Bj%3D1%7D%5E%7Bn%7D%28%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20%5Cwidetilde%7BW_%7Bj%7D%7D%7D%5Ccdot%20sign%28W_%7Bj%7D%29%29&plus;%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20%5Cwidetilde%7BW_%7Bi%7D%7D%7D%20%5Ccdot%20%5Cfrac%7B%5Cpartial%20sign%28W_%7Bi%7D%29%7D%7B%5Cpartial%20W_%7Bi%7D%7D%20%5Ccdot%20%5Calpha)

I will soon write a doc about how to get the equation.

**Comment by jiecaoyu** [comment]
@iridescent303 I have posted the details [here](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/master/notes/notes.pdf). If there is any problem/typo, please let me know. Thanks!

**Comment by iridescent303** [comment]
Oh, I see.Thank you for your reply!

**Comment by caseyanya** [comment]
Thanks a lot for all these works.
This might be the key why I can't even reach a similar result as the original paper in my implementation.
I'll soon try your code.
Thanks.

**Comment by caseyanya** [comment]
Thank you again for your great work.
Finally I have time to look into the code.
There is something that confuse me after reading it, can you give me a more detail explanation?
1.
in the last line of function updateBinaryGradWeight (m.add(m_add).mul(1.0-1.0/s[1]).mul(n))
what is mul(1.0-1.0/s[1]) and also mul(n) doing?
2.
in the imagenet updateBinaryGradWeight
self.target_modules[index].grad.data = self.target_modules[index].grad.data.mul(1e+9)
what does the above line do?

**Comment by caseyanya** [comment]
Also, there is a small mistake in your note.
Eq(11), the last term, it should be partial sign(Wi) / partial (Wi)

**Comment by jiecaoyu** [comment]
@caseyanya Hi, thanks for pointing out the typo!
```mul(1.0-1.0/s[1])``` / ```mul(n)``` / ```mul(1e+9)``` are all strange optimizations the original code has.

1. ```mul(1.0-1.0/s[1])```  is the gradient of mean centering.

2. ```mul(n)```: when we calculate the gradient through Eq(11) or Eq (6), the magnitude of the new gradients will be much lower than the original gradients due to the ```(1/n)``` item. To make them be in the same range, the gradients need to be timed with ```n```.

3. ```mul(1e+9)``` is used to fix the precision.

**Comment by caseyanya** [comment]
@jiecaoyu 
Thanks for your explanation. 
Another question is, I'd like to compare the result on BWN(binary weight network).
I've tried to modify the original code(torch) with the gradient you proposed, but it seems not working.
Since I'm new in torch / pytorch, maybe there's some stupid mistake in it.
Could you please give me some hint?

function updateBinaryGradWeight(convNodes)
   for i =2, #convNodes-1 do
    local n = convNodes[i].weight[1]:nElement()
    local s = convNodes[i].weight:size()
    local m = convNodes[i].weight:norm(1,4):sum(3):sum(2):div(n):expand(s);
    local m_add = convNodes[i].weight:sign():cmul(convNodes[i].gradWeight);
    m[convNodes[i].weight:le(-1)]=0;
    m[convNodes[i].weight:ge(1)]=0;
    m:cmul(convNodes[i].gradWeight);
    m_add:sum(4):sum(3):sum(2):div(n):expand(s);
    m_add:cmul(convNodes[i].weight:sign());
    if opt.optimType == 'sgd' then
       m:mul(n);
    end
    convNodes[i].gradWeight:cmul(m:add(m_add):mul(1-1/s[2]):mul(n))--:cmul(mg)
    end
   if opt.nGPU >1 then
    model:syncParameters()
   end
end

**Comment by jiecaoyu** [comment]
@caseyanya 
one mistake is that you actually time ```m``` twice with ```convNodes[i].gradWeight```
in
```code
m:cmul(convNodes[i].gradWeight);
```
and
```code
convNodes[i].gradWeight:cmul(m:add(m_add):mul(1-1/s[2]):mul(n))--:cmul(mg)
```

**Comment by caseyanya** [comment]
@jiecaoyu 
Thanks a lot. 
I think I've figured out how to make it work, although its convergence seems much slower than your implementation.

**Comment by akhauriyash** [comment]
@jiecaoyu 
Hello, loving the work. However, is there a typo in eqn(12)? Either the last term of the eqn, the partial sign W will have a j subscript with the summation function still on it, or it will have an i subscript. Please check if i am wrong/right. 
Thanks for this awesome repo too!

**Comment by caseyanya** [comment]
@akhauriyash @jiecaoyu 
In my opinion,
For all partial(sign(Wj) / partial(Wi), only i=j is nonzero. 
So, Eq(12) will have an i subscript.

**Comment by kakaxi314** [comment]
In Eq(4), the last term, it should be
$$abs(W_{i})=W_{i} \cdot sign(W_{i})$$
is it right?

**Comment by jiecaoyu** [comment]
@kakaxi314 Yes, I have fixed it. Thanks!

**Comment by kakaxi314** [comment]
@jiecaoyu I try to read and test the code, Because I'm not familiar with the original code of torch,  I love this work very much.
But I still have some questions.

1. The function `meancenterConvParams`  holds part of weights above 0 and the other part below 0, it's a hard constrain to the parameters, is there any explanation for this operation?
2. In BNN, the weights can only be expressed as +1,-1 in the forward path. So clip the weight to [-1,+1] can help optimize. But in XNOR, the weights can be transferred to the network via `\alpha`. Then I wonder if the function `clampConvParams` is beneficial or harmful?
3. The `saved_params` is the mean centering of weights, and finally the gradients will operate on the `saved_params` not the origin weights. So there's no need to add the term `mul(1.0-1.0/s[1])` which is the gradient of mean centering. Or the `saved_params` should be the original weights?
4. You say `mul(1e+9)` is used to fix the precision. I only try to test the mnist and cifar dataset. I don't understand what's the meaning of `fix the precision` and why only ImageNet needs to fix the precision?

Sincerely look forward to discussing with you.

**Comment by jiecaoyu** [comment]
@kakaxi314  Thanks for your questions.
1. Please check my answer [here](https://github.com/allenai/XNOR-Net/issues/25#issuecomment-326888106)

2. As shown in this equation:

![equation](http://latex.codecogs.com/gif.latex?%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20W_%7Bi%7D%7D%3D%5Cfrac%7B1%7D%7Bn%7D%5Ccdot%20sign%28W_%7Bi%7D%29%5Ccdot%20%5Csum_%7Bj%3D1%7D%5E%7Bn%7D%28%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20%5Cwidetilde%7BW_%7Bj%7D%7D%7D%5Ccdot%20sign%28W_%7Bj%7D%29%29&plus;%5Cfrac%7B%5Cpartial%20C%7D%7B%5Cpartial%20%5Cwidetilde%7BW_%7Bi%7D%7D%7D%20%5Ccdot%20%5Cfrac%7B%5Cpartial%20sign%28W_%7Bi%7D%29%7D%7B%5Cpartial%20W_%7Bi%7D%7D%20%5Ccdot%20%5Calpha)

The second element will become zero if ```|W|>1```. The first item, as you said, is calculating the weight gradient via ```\alpha```. However, after training for several epochs, ```\alpha``` will be stable and the gradients toward ```\alpha``` will be small. It may slow down the training. 

3. I think you are right. The right order should be: (1)clamp (2) saved_params (3)meancenterConvParams (4) binarization. In practice, ```mul(1.0-1.0/s[1])``` does not affect the accuracy and can be simply removed.

4. The gradients in the layers closed to the input will be very small. From my understanding, this will be a big problem for AlexNet. In this case, when we are doing the computation of [Adam](http://ruder.io/optimizing-gradient-descent/index.html#adam), we need to calculate the average of past squared gradients. To avoid making the past squared gradients too small, we may need ```mul(1e+9)```.

**Comment by cow8** [comment]
@jiecaoyu thanks for your great work!
but i found some mismatch between the code(util.updateBinaryGradWeight()) and the pdf notes(eq12):
in the code, 

> `m_add = weight.sign().mul(self.target_modules[index].grad.data)`

,where op mul() is element-wise multiplication

in the notes, it's dot product between partial C/ partial W_j and sign(W_j).
![image](https://user-images.githubusercontent.com/16465741/35184722-2c57153c-fe34-11e7-9b78-4513c4ffe05b.png)

i found that the follow ops:
```

if len(s) == 4:
                m_add = m_add.sum(3, keepdim=True)\
                        .sum(2, keepdim=True).sum(1, keepdim=True).div(n).expand(s)

```
Are those ops trying to compensate the error cause by the element-wise mutiplication? Or it's a mistake?

**Comment by zhuyinheng** [comment]
@jiecaoyu thanks for all these work!
But I am confused about the mismatch of the code in  `util.updateBinaryGradWeight()`:
```
if len(s) == 4:
                m_add = m_add.sum(3, keepdim=True)\
                        .sum(2, keepdim=True).sum(1, keepdim=True).div(n).expand(s)
```

and the eq12 in the notes:
![image](https://user-images.githubusercontent.com/16465741/35186008-34f5a47e-fe48-11e7-939d-66a8aa02edf6.png)

I think the eq12 is right, but the implementation doesn't match the eq. and the "right" code i think would be like this:
```
if len(s) == 4:
            m_add = m_add.sum(1,keepdim=True).expand(s)
```
in fact, i'm not sure. Is there any mistakes in my implementation?

**Comment by jiecaoyu** [comment]
@cow8 Hi, thanks for your question!

For convolutional layers, ```m_add``` is a 4D tensor. ```.sum(3, keepdim=True).sum(2, keepdim=True).sum(1, keepdim=True)``` is a naive way to calculate the sum across each filter. Here each filter is a 3D tensor. I think your code consider each filter to be a 1D array.

For example, assuming ```m_add``` has a size of [ N x C x K x K ], performing ```.sum(1, keepdim=True)``` will generate the result with size [ N x 1 x K x K ].

Also, I put ```.div(n)``` after calculating the summation as shown in the equation.

**Comment by zhuyinheng** [comment]
@jiecaoyu thanks!!!
After your explanation, I just realized that I took \W_i as the i^th channel matrix(K*K) of the W(c_in*K*K), but in fact the \W_i is an scalar which is the i^th element in vec(W).
thanks again!!!

**Comment by akhauriyash** [comment]
I actually have a small doube @jiecaoyu .
In this implementation, the updateBinaryGradWeight implements the gradient calculation for dC/dWi correctly. 
However, the BinActive backward function only implements dC/dWi = (dC/dWbi)*d(sign(Wi))/dWi. 

By that, I mean grad_input = grad_output*(d(sign(Wi))/dWi.
    In the backward pass we receive the context object and a Tensor containing
    the gradient of the loss with respect to the output produced during the
    forward pass. We can retrieve cached data from the context object, and must
    compute and return the gradient of the loss with respect to the input to the
    forward function.
Thus, the implementation of gradient calculation must be the same as that which is done in updateBinaryGradWeight, isnt it?

Is this backward calculation (In BinActive) an approximation? I am really confused about this concept.

Any comments on this would be really helpful.

**Comment by mohendra** [comment]
> Hi, I am using:
> 
> ![equation](https://camo.githubusercontent.com/b6992c99badae1fbae6b5906f96a9dcf1a0986c8/687474703a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f253543667261632537422535437061727469616c253230432537442537422535437061727469616c253230575f2537426925374425374425334425354366726163253742312537442537426e25374425354363646f742532307369676e253238575f2537426925374425323925354363646f7425323025354373756d5f2537426a253344312537442535452537426e253744253238253543667261632537422535437061727469616c253230432537442537422535437061727469616c2532302535437769646574696c6465253742575f2537426a25374425374425374425354363646f742532307369676e253238575f2537426a2537442532392532392b253543667261632537422535437061727469616c253230432537442537422535437061727469616c2532302535437769646574696c6465253742575f2537426925374425374425374425323025354363646f74253230253543667261632537422535437061727469616c2532307369676e253238575f253742692537442532392537442537422535437061727469616c253230575f2537426925374425374425323025354363646f74253230253543616c706861)
> 
> I will soon write a doc about how to get the equation.

what is W_j  represents here ?

**Comment by jiecaoyu** [comment]
@mohendra ```j``` is the index for sum. W_i and W_j both refer to the weight.

**Comment by thanhdao** [comment]
Hi Jiecaoyu,
Thank you for your sharing. Could you please tell me how can you have this formula: 
![Screen Shot 2019-04-16 at 10 26 49 PM](https://user-images.githubusercontent.com/3870854/56213367-d8ca2c80-6096-11e9-921e-07afabae5deb.png)

Thank you very much
